### DISCUSSION TITLE
GitHub SSH Handshake failed: knownhosts key mismatch

### DISCUSSION THREAD

**Discussion by jwerre** [fluxcd/flux2#2097] [Q&A] [answered]
I've run into an issue to today where my  deployment rollouts suddenly stopped working. When I checked the logs I get:

```sh
flux get sources git
NAME       	READY	MESSAGE                                                                                                              	REVISION                                       	SUSPENDED
flux-system	False	unable to clone 'ssh://git@github.com/xxxxxx/xxxxxx', error: ssh: handshake failed: knownhosts: key mismatch	master/xxxxxxxxx	False
```

This is the only error in my logs as well. Does anyone know why I'm getting an `ssh: handshake failed` error and how to resolve it?

Here's the output of `flux check`

```sh
► checking prerequisites
✔ Kubernetes 1.19.13-eks-8df270 >=1.19.0-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.11.1
✔ image-automation-controller: deployment ready
► ghcr.io/fluxcd/image-automation-controller:v0.14.0
✔ image-reflector-controller: deployment ready
► ghcr.io/fluxcd/image-reflector-controller:v0.11.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.13.2
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.15.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.15.3
✔ all checks passed
```

**Comment by stefanprodan** [comment, marked as answer]
**Today GitHub has updated its host keys**
https://github.blog/2021-09-01-improving-git-protocol-security-github/

> November 16, 2021 | The ECDSA and Ed25519 host keys will start to be fully usable. GitHub’s DSA host key will no longer be supported.

This impacts those that bootstrapped Flux using ECDSA SSH key format.

To fix the key mismatch error, you have two options:

Update the `known_hosts` in the `flux-system` secret with the `ecdsa-sha2-nistp25` value: 
```
github.com ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEmKSENjQEezOmxkZMy7opKgwFB9nkt5YRrYMjNuG5N87uRgg6CLrbo5wAdT/y6v0mKV0U2w0WZ2YB/++Tpockg=
```

Or rotate the SSH keys with `flux boostrap` like so:
- delete the deploy key secret from your cluster `kubectl -n flux-system delete secret flux-system`
- rerun `flux bootstrap github` with the same arguments as before
- Flux will generate the secret with `ecdsa-sha2` SSH key and Host key

And finally restart source-controller to reset the exponential backoff of cloning retry:

```
kubectl -n flux-system rollout restart deployment/source-controller
```

More details on https://github.com/fluxcd/source-controller/issues/490

**Reply by jwerre** [reply]
Thank you. That solved the issue!

**Reply by aL3891** [reply]
had the exact same issue and this worked for me as well, also a good opportunity to upgrade to the latest flux  :)

**Reply by jvdbrink-netpulse** [reply]
Thank you for providing these solutions. I ran into this issue after a long day yesterday and suspected it was not limited to me. Hoping for a solution I waited till today. When rerunning my google search this discussion showed up :partying_face:, sometimes the stars just align.

**Reply by runningman84** [reply]
The first solution alone does not help... this produces a new error:
```
ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
```

This is the command to update the known_hosts:
```
kubectl patch secret -n flux-system flux-system -p '{"data":{"known_hosts":"Z2l0aHViLmNvbSBlY2RzYS1zaGEyLW5pc3RwMjU2IEFBQUFFMlZqWkhOaExYTm9ZVEl0Ym1semRIQXlOVFlBQUFBSWJtbHpkSEF5TlRZQUFBQkJCRW1LU0VOalFFZXpPbXhrWk15N29wS2d3RkI5bmt0NVlScllNak51RzVOODd1UmdnNkNMcmJvNXdBZFQveTZ2MG1LVjBVMncwV1oyWUIvKytUcG9ja2c9"}}'
```

**Reply by stefanprodan** [reply]
@runningman84 the unable to authenticate error means that the public key is not setup on GitHub or you are using RSA SHA-1. 

What about the 2nd option? Does it work for you? have you update flux to 0.21 or you're using an older version?

**Reply by runningman84** [reply]
The second option works fine but takes a lot of additional effort. We have dozens of clusters therefore a single line solution without any additional unique parameters is always preferred.

**Reply by debu99** [reply]
kubectl scale deploy/source-controller --replicas=0

update the secrets

kubectl scale deploy/source-controller --replicas=1

**Reply by NerdyShawn** [reply]
I had issues with this as well even with updating to latest cli utility. Doing a combination of Stefan's 2nd option in conjunction with scaling down/up the source-controller finally resolved it.

1. kubectl scale deploy/source-controller --replicas=0
2. kubectl -n flux-system delete secret flux-system
3. kubectl scale deploy/source-controller --replicas=1
4. then my flux bootstrap

**Reply by mohsinhijazee** [reply]
May be something like 
`flux check hosts --reset` or similar without requiring the whole bootstrap process which requires bootstrap token. 
could be useful in future.

**Reply by jrschumacher** [reply]
I was getting `error GitRepository/flux-system.flux-system - failed to checkout and determine revision`[^error] error and `flux uninstall --keep-namespace` did not resolve the issue.

By uninstalling flux and removing the key I was able to bootstrap again and get it working:

```
flux uninstall --keep-namespace
kubectl -n flux-system delete secret flux-system
# Run bootstrap agian
```

[^error]: `GitRepository/flux-system.flux-system - failed to checkout and determine revision: unable to clone 'ssh://git@github.com/xxx/gitops': ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain failed to checkout and determine revision`

**Comment by pantelis-karamolegkos** [comment]
I assume this affects `flux2` since it uses `go-git` under the hood right? (asking since we also faced some issues with programs using `go-git` but no relevant problems with plain `git clone`

**Reply by stefanprodan** [reply]
Yes, Flux the CLI uses `go-git` but Flux source-controller allows users to chose between `go-git` and `libgit2`, both implementations were affected by this.
